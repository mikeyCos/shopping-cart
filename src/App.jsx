import { BrowserRouter } from "react-router-dom";
import Routes from "./routes/Routes";
import styles from "./styles/App.module.css";
import { CartProvider } from "./components/Cart";

const App = () => {
  return (
    <div id={styles.app}>
      <CartProvider>
        <BrowserRouter>
          <Routes />
        </BrowserRouter>
      </CartProvider>
    </div>
  );
};

export default App;

/*
Navigation bar
  - Logo Link
    - Goes to root
  - Home Link
    - Goes to root
  - Shop Link
    - Goes to shop page
  - Shopping cart Link
    - If cart has an item, show a number near the cart icon
    - Goes to a checkout page
Two pages
  - Home
    - Contain filler content
  - Shop
    - Make use of https://fakestoreapi.com/
    - Individual card elements for each product
      - Input quantity (type number)
        - (OPTIONAL), create custom increment and decrement buttons
        - Need to hide the inner-spin-button
        - https://stackoverflow.com/questions/75879418/how-to-remove-arrows-in-input-type-number-inside-the-input
      - Title
      - "Add To Cart" button
- BrowserRouter does not support the react-router-dom data APIs
*/
