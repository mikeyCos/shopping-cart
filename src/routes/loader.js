import { defer } from "react-router-dom";
import parseProducts from "../utilities/parseProducts";

const loader = async () => {
  const data = await Promise.all([
    fetch("https://fakestoreapi.com/products/categories", { mode: "cors" }),
    fetch("https://fakestoreapi.com/products", { mode: "cors" }),
  ])
    .then(([resCategories, resProducts]) => {
      // How to mock response status codes >= 400?
      if (resCategories.status >= 400) throw new Error("ERROR");
      return Promise.all([resCategories.json(), resProducts.json()]);
    })
    .then(([dataCategories, dataProducts]) => {
      const parsedProducts = parseProducts(dataProducts);
      return {
        categories: ["all", ...dataCategories],
        products: parsedProducts,
      };
    })
    .catch((error) => error);
  return data;
};

// const loader = async () => {
//   const data = Promise.all([
//     fetch("https://fakestoreapi.com/products/categories", { mode: "cors" }),
//     fetch("https://fakestoreapi.com/products", { mode: "cors" }),
//   ])
//     .then(([resCategories, resProducts]) => {
//       // How to mock response status codes >= 400?
//       if (resCategories.status >= 400) throw new Error("ERROR");
//       return Promise.all([resCategories.json(), resProducts.json()]);
//     })
//     .then(([dataCategories, dataProducts]) => {
//       const parsedProducts = parseProducts(dataProducts);
//       return {
//         categories: ["all", ...dataCategories],
//         products: parsedProducts,
//       };
//     })
//     .catch((error) => error);
//   return defer({ data });
// };

export default loader;
