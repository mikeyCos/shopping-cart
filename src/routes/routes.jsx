import { useLocation, useRoutes } from "react-router-dom";
import loader from "./loader";
import BasicLayout from "../layouts/BasicLayout";
import Modal from "../components/Modal";
import Product from "../components/Product";
import ErrorPage from "../components/ErrorPage";
import Home from "../components/Home";
import Shop from "../components/Shop";
import Cart from "../components/Cart";
import Category from "../components/Category";
import { Suspense } from "react";

const routes = [
  {
    element: <BasicLayout />,
    // element: (
    //   <Suspense fallback={<h1>LOADING...</h1>}>
    //     <BasicLayout />
    //   </Suspense>
    // ),
    errorElement: <ErrorPage />,
    id: "root",
    loader: loader,
    children: [
      {
        path: "/home?",
        element: <Home />,
      },
      {
        path: "/shop",
        element: <Shop />,
        children: [
          {
            path: "category?/:category?",
            element: <Category />,
            children: [
              {
                path: "product/view/:product",
                element: null,
              },
            ],
          },
        ],
      },
      {
        path: "/cart",
        element: <Cart />,
      },
    ],
  },
];

export default routes;
