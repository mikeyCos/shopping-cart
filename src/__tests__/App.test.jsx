import { beforeEach, describe, expect, it, vi } from "vitest";
import {
  render,
  screen,
  waitForElementToBeRemoved,
} from "@testing-library/react";
import { createBrowserRouter } from "react-router-dom";
import App from "../App";
import routes from "../routes/routes";

beforeEach(() => {
  window.scrollTo = vi.fn();
});

describe("App component", () => {
  it("The App component matches snapshot", async () => {
    const router = createBrowserRouter(routes);
    const { container } = render(<App router={router} />);

    await waitForElementToBeRemoved(() => screen.getByTitle("Loading"));
    expect(container).toMatchSnapshot();
  });

  it("Renders splash screen on load and while API request is in progress", () => {
    const router = createBrowserRouter(routes, { initialEntries: ["/"] });
    render(<App router={router} />);
    const loading = screen.getByTitle("Loading");
    expect(loading).toBeInTheDocument();
  });
});
