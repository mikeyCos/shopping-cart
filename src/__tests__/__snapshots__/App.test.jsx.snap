// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App component > The App component is rendered 1`] = `
<div>
  <div
    id="_app_61351a"
  />
</div>
`;
